src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -321,35 +337,102 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
       } else t
     }
 
+    /** Annotations ::= Annotation* */
     def annotations(): List[Tree] = {
-      //var annots = new ListBuffer[Tree]
+      val annots = new ListBuffer[Tree]
       while (in.token == AT) {
         in.nextToken()
-        annotation()
+        annots += annotation()
       }
-      List() // don't pass on annotations for now
+      annots.toList
     }
 
-    /** Annotation ::= TypeName [`(` AnnotationArgument {`,` AnnotationArgument} `)`]
+    /** Annotation              ::= `@` (NormalAnnotation | MarkerAnnotation | SingleElementAnnotation)
+     *  NormalAnnotation        ::= Ident `(` ElementValuePair* `)`
+     *  MarkerAnnotation        ::= Ident
+     *  SingleElementAnnotation ::= Ident `(` ElementValue `)`
+     *
+     *  ElementValuePair        ::= Ident `=` ElementValue
+     *  ElementValue            ::= ConditionalExpression | Annotation | ElementValueArray
+     *  ConditionalExpression   ::=
+     *  ElementValueArray       ::= `{` (ElementValue `,`)* `,`?  }
      */
-    def annotation() {
-      qualId()
-      if (in.token == LPAREN) { skipAhead(); accept(RPAREN) }
-      else if (in.token == LBRACE) { skipAhead(); accept(RBRACE) }
+    def annotation(): Tree = {
+      val t = convertToTypeId(qualId())
+      val args: ListBuffer[List[Tree]] = ListBuffer.empty
+      if (in.token == LPAREN) { // NormalAnnotation
+        in.nextToken()
+        if (in.token == RPAREN) { // NormalAnnotation without any ElementValuePair, e. g. @Foo()
+          // Nothing to do
+        } else if (in.token == LBRACE) { // SingleElementAnnotation where the element is an array, e. g. @Foo({1,2,3})
+          args += List(elementValueArray)
+        } else if (in.token == IDENTIFIER && in.lookaheadToken == EQUALS) { // NormalAnnotation with ElementValuePairs, e. g. @Foo(bar = BAZ, ber = BEZ)
+          args ++= repsep(elementValuePair, COMMA).map(annot => List(annot))
+        } else { // SingleElementAnnotation
+          args += List(elementValue())
+        }
+        accept(RPAREN)
+      } else { // MarkerAnnotation, e. g. @Foo
+        // Nothing to do
+      }
+      New(t, args.toList)
+    }
+
+    def elementValuePair(): Tree = {
+      val elementName = Ident(ident())
+      accept(EQUALS)
+      val elementValue = qualIdOrClassLiteral()
+      AssignOrNamedArg(elementName, elementValue)
+    }
+
+    def elementValue(): Tree = in.token match {
+      case AT =>                       annotation()
+      case LBRACE =>                   elementValueArray()
+      case IDENTIFIER =>               qualIdOrClassLiteral()
+      case _ if isLiteral(in.token) => literal()
```

Comment:
Echoing my earlier comment: The java spec allows any constant expression as annotation value, not just literals, see https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.7.1 / https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.28

See also https://issues.scala-lang.org/browse/SI-5333